### PR TITLE
crimson/onode-staged-tree: integrate interruptible future

### PR DESCRIPTION
**Pull request by liu-chunmei**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by tchaikov — APPROVED** [review]
just a few nits.

File: src/tools/crimson/perf_staged_fltree.cc
Comment (liu-chunmei) [reply src/tools/crimson/perf_staged_fltree.cc]: I removed the macro definition from two test files, but another one (test_stages_fltree.cc) has several tens of places to change, so I think keep the macro will let the code looks clean and read easily.

**Comment by tchaikov** [comment]
@cyx1231st mind taking a look?

**Comment by cyx1231st** [comment]
Yeah, still looking.

File: src/test/crimson/seastore/onode_tree/test_staged_fltree.cc
Comment (cyx1231st) [inline src/test/crimson/seastore/onode_tree/test_staged_fltree.cc:49]: How about replacing all INTR_WITH_PARAM by INTR_R and remove it?

File: src/test/crimson/seastore/onode_tree/test_staged_fltree.cc
Comment (cyx1231st) [inline src/test/crimson/seastore/onode_tree/test_staged_fltree.cc:1196]: nit, wrong indent

File: src/crimson/os/seastore/onode_manager/staged-fltree/node.cc
Comment (liu-chunmei) [reply src/crimson/os/seastore/onode_manager/staged-fltree/node.cc:403]: other tree lba, omap mkfs not use interruptible future, so I keep onode mkfs also not use interruptible future.

File: src/test/crimson/seastore/onode_tree/test_staged_fltree.cc
Comment (liu-chunmei) [reply src/test/crimson/seastore/onode_tree/test_staged_fltree.cc:41]: seems can't.  here pass c, and use c.t in with_trans_intr. and fun also use c,

**Comment by tchaikov** [comment]
jenkins test make check

**Review by athanatos — COMMENTED** [review]
LGTM once @cyx1231st approves.

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree.h
Comment (cyx1231st) [inline src/crimson/os/seastore/onode_manager/staged-fltree/tree.h:124]: Why remove this assert?

File: src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h
Comment (cyx1231st) [inline src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h:116]: Remove line 116 ~ 121

**Comment by cyx1231st** [comment]
Seems crimson-OSD will crush with this PR when creating test pool.
```
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - Cache::prepare_record(0x600000173000): enter
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - Cache::prepare_record(0x600000173000): read_set validated
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - TransactionManager::submit_transaction_direct(0x600000173000): about to submit to journal
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - write_record, mdlength 4096, dlength 0, target 118784
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - segment_write to segment 5 at offset 118784, physical offset 335671296, len 4096
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - block: do_writev offset 335671296 len 4096
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - ~btree_range_pin_t: removing btree_range_pin_t(begin=100696064, end=100700160, depth=0, extent=0x602b9f061a40)
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - remove_pin: btree_range_pin_t(begin=100696064, end=100700160, depth=0, extent=0x602b9f061a40)
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - check_parent: releasing parent btree_range_pin_t(begin=0, end=18446744073709551615, depth=1, extent=0x6000001bd500)
DEBUG 2021-07-30 11:01:05,545 [shard 0] osd - pg_advance_map(id=5, detail=PGAdvanceMap(pg=1.4 from=5 to=5 do_init)): complete
DEBUG 2021-07-30 11:01:05,545 [shard 0] seastore - "repeat_eagain": hit eagain, restarting
Segmentation fault on shard 0.
Backtrace:
 0# 0x000055DC898963C6 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 1# FatalSignal::signaled(int, siginfo_t const&) in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 2# FatalSignal::install_oneshot_signal_handler<11>()::{lambda(int, siginfo_t*, void*)#1}::_FUN(int, siginfo_t*, void*) in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 3# 0x00007F574FC063C0 in /lib/x86_64-linux-gnu/libpthread.so.0
 4# crimson::os::seastore::CachedExtent::~CachedExtent() in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 5# crimson::os::seastore::onode::SeastoreNodeExtent::~SeastoreNodeExtent() in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 6# std::__cxx11::_List_base<boost::intrusive_ptr<crimson::os::seastore::CachedExtent>, std::allocator<boost::intrusive_ptr<crimson::os::seastore::CachedExtent> > >::_M_clear() in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 7# 0x000055DC8A278791 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 8# 0x000055DC8A279A3C in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 9# 0x000055DC8A279FB3 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
10# 0x000055DC8A27A02D in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
11# 0x000055DC8A733B03 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
12# 0x000055DC8A73CA30 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
13# 0x000055DC8A76287C in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
14# 0x000055DC8A6E69C9 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
15# 0x000055DC8A6E7159 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
16# main in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
17# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
18# _start in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
```

**Comment by cyx1231st** [comment]
Hi @tchaikov @rzarzynski Do you have any clue why stack cannot be parsed? For example:
```
 7# 0x000055DC8A278791 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 8# 0x000055DC8A279A3C in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
 9# 0x000055DC8A279FB3 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
10# 0x000055DC8A27A02D in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
11# 0x000055DC8A733B03 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
12# 0x000055DC8A73CA30 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
13# 0x000055DC8A76287C in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
14# 0x000055DC8A6E69C9 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
15# 0x000055DC8A6E7159 in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
```
Seems line number is also not available:
```
4# crimson::os::seastore::CachedExtent::~CachedExtent() in /root/ceph/obj-x86_64-linux-gnu-debug/bin/crimson-osd
```

**Comment by cyx1231st** [comment]
I'm looking into the issue.

**Comment by cyx1231st** [comment]
Sorry fetched a wrong PR to test. This PR works good.

**Comment by cyx1231st** [comment]
Changeset: crimson/onode-staged-tree: misc fixes to integrate interruptive-future

**Review by cyx1231st — APPROVED** [review]
LGTM

**Pull request merged**
